qutebrowser/browser/urlmarks.py

```diff
@@ -0,0 +1,296 @@
+        message.ask_async(
+            win_id, "Add quickmark:", usertypes.PromptMode.text,
+            functools.partial(self.quickmark_add, win_id, urlstr))
+
+    @cmdutils.register(instance='quickmark-manager', win_id='win_id')
+    def quickmark_add(self, win_id, url, name):
+        """Add a new quickmark.
+
+        Args:
+            win_id: The window ID to display the errors in.
+            url: The url to add as quickmark.
+            name: The name for the new quickmark.
+        """
+        # We don't raise cmdexc.CommandError here as this can be called async
+        # via prompt_save.
+        if not name:
+            message.error(win_id, "Can't set mark with empty name!")
+            return
+        if not url:
+            message.error(win_id, "Can't set mark with empty URL!")
+            return
+
+        def set_mark():
+            """Really set the quickmark."""
+            self.marks[name] = url
+            self.changed.emit()
+            self.added.emit(name, url)
+
+        if name in self.marks:
+            message.confirm_async(
+                win_id, "Override existing quickmark?", set_mark, default=True)
+        else:
+            set_mark()
+
+    @cmdutils.register(instance='quickmark-manager', maxsplit=0,
+                       completion=[usertypes.Completion.quickmark_by_name])
+    def quickmark_del(self, name):
+        """Delete a quickmark.
+
+        Args:
+            name: The name of the quickmark to delete.
+        """
+        try:
+            self.delete(name)
+        except KeyError:
+            raise cmdexc.CommandError("Quickmark '{}' not found!".format(name))
+
+    def get(self, name):
+        """Get the URL of the quickmark named name as a QUrl."""
+        if name not in self.marks:
+            raise DoesNotExistError(
+                "Quickmark '{}' does not exist!".format(name))
+        urlstr = self.marks[name]
+        try:
+            url = urlutils.fuzzy_url(urlstr, do_search=False)
+        except urlutils.FuzzyUrlError as e:
+            raise InvalidUrlError(
+                "Invalid URL for quickmark {}: {}".format(name, str(e)))
+        return url
+
+
+class BookmarkManager(UrlMarkManager):
+
+    """Manager for bookmarks.
+
+    The primary key for bookmarks is their *url*, this means:
+
+        - self.marks maps URLs to titles.
+        - changed gets emitted with the URL as first argument and the title as
+          second argument.
+        - removed gets emitted with the URL as argument.
+    """
+
+    def _init_lineparser(self):
+        bookmarks_directory = os.path.join(standarddir.config(), 'bookmarks')
+        if not os.path.isdir(bookmarks_directory):
+            os.makedirs(bookmarks_directory)
+        self._lineparser = lineparser.LineParser(
+            standarddir.config(), 'bookmarks/urls', parent=self)
+
+    def _init_savemanager(self, save_manager):
+        filename = os.path.join(standarddir.config(), 'bookmarks/urls')
```

Comment:
I could have put the `'bookmarks/urls'` (from the two previous methods) in a class constant.